src/main/scala/it/scalalearn/calculator/Parser.scala

```diff
@@ -0,0 +1,191 @@
+      if (t == PLUS || t == DASH) {
+        val (tokensAfterRight, rightRoot, newParenLevelRight) =
+          parseFactor(tokensAfterLeft.tail, EmptyNode(), newParenLevelLeft)
+        parseTerm(tokensAfterRight, TermNode(t, leftRoot, rightRoot), newParenLevelRight)
+      } else (tokensAfterLeft, leftRoot, newParenLevelLeft)
+    }
+  }
+
+  /**
+   * Parses a factor (product or quotient) expression, including left associativity
+   *
+   * @param  tokens     tokens remaining to be parsed in this expression
+   * @param  leftRootIn if this expression is the second argument in an infix (binary) operator,
+   *                    represents the node at the root of the parse tree of the first argument
+   * @param  parenLevel a List containing the as-yet unclosed parentheses at this level
+   * @return            a tuple containing:
+   *                      1. a list of tokens remaining after parsing this expression
+   *                      2. the root node of the parse tree generated by this expression
+   *                      3. a list of the parenthesis depth after parsing this expression
+   */
+  @tailrec
+  private def parseFactor(tokens: List[Token],
+                          leftRootIn: ParseNode,
+                          parenLevel: List[Token]): (List[Token], ParseNode, List[Token]) = {
+
+    val (tokensAfterLeft, leftRoot, newParenLevelLeft) =
+      if (leftRootIn.isEmpty) parseSign(tokens, parenLevel)
+      else (tokens, leftRootIn, parenLevel)
+
+    if (tokensAfterLeft.isEmpty) (tokensAfterLeft, leftRoot, newParenLevelLeft)
+    else {
+      val t = tokensAfterLeft.head
+      if (t == STAR || t == SLASH) {
+        val (tokensAfterRight, rightRoot, newParenLevelRight) =
+          parseSign(tokensAfterLeft.tail, newParenLevelLeft)
+        parseFactor(tokensAfterRight, FactorNode(t, leftRoot, rightRoot), newParenLevelRight)
+      } else (tokensAfterLeft, leftRoot, newParenLevelLeft)
+    }
+  }
+
+  /**
+   * Parses unary (sign) expression, including right associativity
+   *
+   * @param  tokens     tokens remaining to be parsed in this expression
+   * @param  parenLevel a List containing the as-yet unclosed parentheses at this level
+   * @return            a tuple containing:
+   *                      1. a list of tokens remaining after parsing this expression
+   *                      2. the root node of the parse tree generated by this expression
+   *                      3. a list of the parenthesis depth after parsing this expression
+   */
+  private def parseSign(tokens: List[Token], parenLevel: List[Token]): (List[Token], ParseNode, List[Token]) = {
+    if (tokens.isEmpty) throw new ParserException("expression terminated where a value was expected")
+
+    val t = tokens.head
+    if (t == DASH) {
+      val (remainingTokens, number, newParenLevel) = parseSign(tokens.tail, parenLevel)
+      (remainingTokens, SignNode(t, number), newParenLevel)
+    }
+    else parseNumber(tokens, parenLevel)
```

Comment:
Pattern match would help here in readability

```suggestion
    tokens match {
      case Nil => throw new ParserException("expression terminated where a value was expected")
      case DASH :: rest =>
              val (remainingTokens, number, newParenLevel) = parseSign(rest, parenLevel)
              (remainingTokens, SignNode(t, number), newParenLevel)
      case _ => parseNumber(tokens, parenLevel)
```

Reply:
That `DASH :: rest` pattern is very useful. I'll see if I can use the same idea elsewhere in that file.